doc/developer/design/20210630_linearizability.md

```diff
@@ -0,0 +1,108 @@
+
+## Goals
+
+We want to provide linearizability in the SQL layer (the coordinator) as defined by [Jepsen] which says that
+
+> every operation appears to take place atomically, in some order, consistent with the real-time ordering of those operations.
+
+The goal is to define the requirements of the coordinator and assumptions it makes about other parts of the system to provide this.
+
+This is limited to transactions (or single statements) in the SQL layer where materialize is allowed to choose the timestamp for an operation.
+Statements can use `AS OF` to specify the timestamp of an operation.
+
+Since dataflow provides serializability, the addition of linearizability would also provide strict serializability over the SQL layer (mostly, see Catalog discussion below).
+
+[Jepsen]: http://jepsen.io/consistency/models/linearizable
+
+## Non-Goals
+
+Linearizability across the entire system.
+This may come in a later document.
+
+Verifying assumptions about other parts of the system.
+
+## Description
+
+The coordinator assumes that if dataflow returns a result at a timestamp that it will always return the same result at that timestamp, even after restart.
+We are aware that this is not the case today, but document this assumption to draw clear lines of responsibility.
+For example, we have volatile sources whose data changes after restart,
+and non-volatile sources that may have different timestamp bindings after restart because they return results before timestamps are durably persisted.
+
+To provide linearizability, when a SQL transaction (which can be a single statement) needs to generate a timestamp (either at commit time for INSERT or the first SELECT or TAIL in a transaction),
+that timestamp must be >= any other timestamp generated by any other transaction timestamp that is generated after it in real time.
+This will ensure that should some SQL result be observed by some user,
+that user can tell some other user about its results,
+and the second user is guaranteed to observe those results (unless they've changed in the meantime).
+
+However, we cannot have a single timestamp number that we guarantee always goes up, because timestamps do not always have the same meaning.
+We have solved the different meaning problem with timelines, which allow joining different sources only if their timestamp number means the same thing.
+For linearizability, we need to have a timestamp number for each timeline.
+
+Add a `HashMap<Timeline, Timestamp>` to track the highest timestamp for a given timeline.
+`sequence_peek`, `sequence_tail`, and `sequence_end_transaction` are the three places timestamps are generated for transactions.
+These three functions will call some new helper function that will either advance or return the most recent timestamp in the HashMap.
+
+Tables are special in that they already provide linearizability.
```

Comment:
The timestamp functions used by tables are used by the three sequence functions from the previous paragraph, and so are included in the HashMap. But I think it's easier to just remove this paragraph because it's not important to this design.